src/pwm.rs

```diff
@@ -0,0 +1,186 @@
+    ///
+    /// ## Important:
+    /// It is **important** to make sure that the provided **device is already initialized** by using [`riot_sys::pwm_init`](https://rustdoc.etonomy.org/riot_sys/fn.pwm_init.html).
+    /// Using the returned device otherwise results in **undefined behavior**!
+    ///
+    /// Also note, that the given frequency is only to be used by [`embedded_hal::Pwm::get_period`] or [`PWMDevice::get_frequency`].
+    /// Just setting this to `x` will only result in those two functions returning `x` but will have no other impact, if
+    /// for example the given [`pwm_t`] is initialized by a board and the actually set frequency is unknown.
+    ///
+    pub unsafe fn new_without_init(dev: pwm_t, frequency: Hz) -> Self {
+        PWMDevice {
+            dev,
+            channel_duty_values: [0; CHANNELS],
+            frequency,
+        }
+    }
+
+    /// Initializes the [`PWMDevice`] with the given frequency in Hz and resolution of a period/duty cycle.
+    ///
+    /// If the given frequency and resolution values can not be acchieved on the current device, the
+    /// resolution will be kept the same while the frequency will be lowered until the device can handle the combination.
+    /// The resulting frequency is written into the [`PWMDevice`].
+    ///
+    /// Uses a [`Result`] in anticipation of a more usefull error handling in the future.
+    ///
+    /// Returns if the initialization was a success
+    fn init(&mut self, mode: PWMMode, frequency: Hz, resolution: u16) -> Result<(), ()> {
+        let err = unsafe { riot_sys::pwm_init(self.dev, mode.to_c(), frequency, resolution) };
+
+        match err {
+            0 => Err(()),
+            freq => {
+                // Set frequency in the device
+                self.frequency = freq;
+                Ok(())
+            }
+        }
+    }
+
+    /// Returns the number of available channels for this device
+    pub fn channels(&self) -> u8 {
+        unsafe { riot_sys::pwm_channels(self.dev).min(CHANNELS as u8) }
+    }
+
+    /// Stops PWM generation on this device
+    pub fn power_off(&mut self) {
+        unsafe { riot_sys::pwm_poweroff(self.dev) }
+    }
+
+    /// Resumes PWM generation after power_off on this device
+    pub fn power_on(&mut self) {
+        unsafe {
+            riot_sys::pwm_poweron(self.dev);
+        }
+    }
+
+    /// Sets the duty-cycle for the given channel
+    ///
+    /// value: `0: 0%, u16::MAX: 100%` duty_cycle
+    pub fn set(&mut self, channel: u8, value: u16) {
+        unsafe {
+            riot_sys::pwm_set(self.dev, channel, value);
+        }
+
+        let channel = channel as usize;
+        // Ignore if entry does not exists because
+        // only embedded_hal interface cares about those values
+        // and the implementation already checks for this
+        if channel < (CHANNELS) {
```

Comment:
Over at `.channels()` we treat the PWMDevice as clipped to `CHANNELS` channels. For consistency, we should do this here as well, and err out without setting anything ... maybe even panic when using a channel that doesn't actually exist?

(Not sure on the best direction here).

Reply:
Oh i just realized that there is a bit of a Problem. In the comments for `PWMDevice` i say its o.k. to just set `CHANNELS=0` if `embedded_hal` is not used but in `channels()` I act as if `CHANNELS` would always be the count of available channels, which might not be the case!

I would try to change the api a bit:

* Remove `CHANNELS`
* Add `use_embedded_hal: bool` to `new/new_without_init`
* Setup the duty cache with `riot_sys::pwm_channels` if `use_embedded_hal`

In addition I could insert a check using `channels` to `set` as you suggested to instead of RIOT failing at `pwm_set` give a "clean" Rust `panic`.

Reply:
I don't quite get the proposal: How would, with CHANNELS removed, the values be stored for the getters?

Reply:
The idea is to determine the length for the duty cache with `riot_sys::pwm_channels` in the `new` method instead of taking this length in via `CHANNELS`. We still have the cache array but determine its size in a different way.  

This wastes a little bit of memory if not all of the available channels are used but might be worth the simpler API.

Reply:
If a big array gets allocated all the time I think we should use it.

But how about we take the same approach as with the panicing functions, and offer CHANNELS as a way out?

Channel size report etc. would all work on the actual (RIOT-reported) number of channels, but getting the set PWM value would panic if CHANNELS is not big enough.

Thus, a user can decide whether to just take the zero-cost abstraction and forfeit the use of the accessor, or or get the fuller trait implementation at the cost of having [u16; CHANNELS] stored. (And thanks to monomorphization, all the "if channel > CHANNELS" checks, eg. for writing to the array in the setter, would be optimized out).

Reply:
(My original Idea was to allocate the array only if the user consented by setting `cache_duty`, but checks (`if duty_cache` in that case) would not be optimized out that way)

The only little downside i see with that is, that the user has not the opportunity to dynamically determine the available channels when setting `CHANNELS` at compile time. So he has to look this information up for his Board and possibly change his code when switching boards instead of being able to dynamically act on different available channel counts. (Not sure if there are any use-cases for that though)

Reply:
If ever RIOT's hardware setup becomes more descriptive, we may have a `pwm_channels()` function that can be `const`. Right now, it is very much not.

Requiring the user to give some number in the type seems to me to be the only alternative to unconditionally allocating for a large number of channels. I think that's a reasonable requirement for the time being. It may be prudent to allow CHANNELS to be larger than the actual number of channels, so that applications that don't know their hardware well can just do `let pwm_devices [_; _]  = (0..16).map(PwmDevice::<16>::new).collect();`. (Sadly, AFAIK collect can't do that yet in this very syntax).

Reply:
Ah i had forgotten that we cant just do dynamic allocation. Then i don't see any alternative to `CHANNELS` either at the moment except for bringing the problem to the user by introducing a `struct Channel` which caches that information and is being provided to `set_duty` But then you could have two instances for the same channel with different duty_caches which seems just a bit ugly.

I would go with what you described above:

> ...
> 
> But how about we take the same approach as with the panicing functions, and offer CHANNELS as a way out?
> 
> Channel size report etc. would all work on the actual (RIOT-reported) number of channels, but getting the set PWM value would panic if CHANNELS is not big enough.
> 
> ...

Reply:
While making those changes i was wondering if we want to make a check on `channels()` in `set_duty`. The original check against `CHANNELS` has to go since, we want `CHANNELS` to be possibly `0`. I am not sure whats best practice there, RIOT will probably fail in `riot_sys::pwm_set` in some way but a Rust panic might be easier to debug but then  that will introduce checks to every `set_duty` call.

Reply:
A PWM with different instances is nothing I think we should worry about. RIOT has no ownership concept and is free-for-all. When creating a Rust instance for a peripheral, the user states that they intend to use this exclusively. We can't allow any memory-unsafe effects when this is violated (for otherwise the constructors would need to be unsafe), but effects like getters getting confused sounds OK to me.

What's the problem with `set_duty`? It'd forward the write to RIOT, and only write to the cache if CHANNELS is large enough. (In performance critical situations that won't be a check at all, because there the CHANNELS would be monomorphized to 0, and every number is >= 0).

Reply:
What bothered me was that in `set_duty` `riot_sys::pwm_set` gets called directly with the provided `channel: u8`. So if this is not a valid channel the underlying implementation of pwm might crash, since no error is returned. I was wondering if it might make sense to catch that case beforehand in `set_duty` (as seen here https://github.com/RIOT-OS/rust-riot-wrappers/pull/38/commits/a8a9ea0dbf84b9d8a2628025e87e272d5b178310) to allow for a possibly more understandable, Rust based error message. 

If we decide we don't like the check after all I'll just remove it.

Reply:
I've sampled three PWM implementations, and indeed they show the worst possible variation for nonexisting channel access:

* cpu/nrf52/periph/pwm.c silently ignores such writes.
* cpu/lpc23xx/periph/pwm.c places an assert, but C assertions are disabled, it silently ignores writes.
* cpu/esp8266/periph/pwm.c places an assert, and if C assertions are disabled, it happily accesses the register array out-of-bounds, which may be a write to a completely unrelated register. Kaboom.

The (unstated?) goal of riot-wrappers is to be safe, zero-cost, and ergonomic / idiomatic. We can't reach this all the time, but we can try. The current solution is safe (as safe as we can be relying on RIOT's guarantees) but does a bounds check of unknown cost on each set.

An alternative that is a bit less easy-to-use but more performant would be to check the channel number against the const channel count at construction time -- we wouldn't get the full Rust typestate, but at least we could panic (or no-op, IMO either is fine) more cheaply.

I'd say that the current solution is good enough, especially given that it'll be up for some redesign with embedded-hal 1.1. If you want to explore alternatives, be my guest, but I'm fine with how it is right now.

Reply:
In https://github.com/RIOT-OS/rust-riot-wrappers/pull/38/commits/f89082d04156f6df68e9c80c5a90202f6ce1a9c2 I tried an idea to use a separate `PWMChannel` struct which only holds a `u8`. `set`/`set_duty` now take this `struct`. Which can only be created from the `PWMDevice::get_channel` where we check against `channels()`. This would trade the check costs for a bit of memory. 

What are your thoughts on that approach?